File: android/app/src/main/java/com/zagir/splitty/ui/components/DesignSystem.kt
package com.zagir.splitty.ui.components

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.zagir.splitty.core.model.CurrencySum
import com.zagir.splitty.core.money.money
import com.zagir.splitty.ui.theme.Splitty
import kotlin.math.abs

// Компоненты дизайн-системы — порт ios/Splitty/Core/DesignSystem.swift.
// Все секции экранов — SurfaceCard на фоне Splitty.colors.bg; деньги —
// ТОЛЬКО через MoneyText/MoneyTotalsText.

// MARK: Карточка-поверхность

/**
 * Премиум-карточка: фон `surface`, скругление 20dp; светлая тема — мягкая
 * тень, тёмная — hairline-бордер без тени. Аналог iOS `.surfaceCard()`.
 */
@Composable
fun SurfaceCard(
    modifier: Modifier = Modifier,
    padding: Dp = 16.dp,
    content: @Composable ColumnScope.() -> Unit,
) {
    val colors = Splitty.colors
    val shape = RoundedCornerShape(20.dp)
    val cardModifier = if (colors.isDark) {
        modifier
            .clip(shape)
            .background(colors.surface)
            .border(1.dp, colors.hairline, shape)
    } else {
        modifier
            .shadow(
                elevation = 6.dp,
                shape = shape,
                ambientColor = Color.Black.copy(alpha = 0.06f),
                spotColor = Color.Black.copy(alpha = 0.06f),
            )
            .clip(shape)
            .background(colors.surface)
    }
    Column(modifier = cardModifier.padding(padding), content = content)
}

// MARK: Кнопки

/**
 * Основной CTA: pill во всю ширину, высота 54dp, фон `accent`, белый
 * semibold-текст, pressed — scale 0.98. Аналог iOS `.primaryPill`.
 */
@Composable
fun PrimaryPillButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
) {
    val colors = Splitty.colors
    val reduceMotion = rememberReduceMotion()
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()
    val scale by animateFloatAsState(
        // reduce motion: без «прожатия» — как iOS `!reduceMotion && isPressed`.
        targetValue = if (pressed && !reduceMotion) 0.98f else 1f,
        animationSpec = spring(stiffness = Spring.StiffnessMedium),
        label = "pillScale",
    )
    Button(
        onClick = onClick,
        modifier = modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = 54.dp)
            .scale(scale),
        enabled = enabled,
        shape = CircleShape,
        colors = ButtonDefaults.buttonColors(
            containerColor = if (pressed) colors.accentPressed else colors.accent,
            contentColor = Color.White,
            disabledContainerColor = colors.accent.copy(alpha = 0.45f),
            disabledContentColor = Color.White,
        ),
        interactionSource = interactionSource,
    ) {
        Text(text = text, fontSize = 17.sp, fontWeight = FontWeight.SemiBold)
    }
}

/**
 * Вторичная кнопка-чип: мягкая серая pill; [isSelected] — акцентная заливка
 * (для выбора группы/участника, фильтров). Аналог iOS `.softChip(isSelected:)`.
 */
@Composable
fun SoftChip(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    isSelected: Boolean = false,
) {
    val colors = Splitty.colors
    val reduceMotion = rememberReduceMotion()
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()
    val scale by animateFloatAsState(
        targetValue = if (pressed && !reduceMotion) 0.98f else 1f,
        animationSpec = spring(stiffness = Spring.StiffnessMedium),
        label = "chipScale",
    )
    val background = when {
        isSelected -> colors.accent.copy(alpha = 0.14f)
        pressed -> colors.ink.copy(alpha = 0.12f)
        else -> colors.ink.copy(alpha = 0.06f)
    }
    Box(
        modifier = modifier
            .scale(scale)
            .clip(CircleShape)
            .background(background)
            .clickable(
                interactionSource = interactionSource,
                indication = null,
                onClick = onClick,
            )
            .padding(horizontal = 16.dp, vertical = 10.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = text,
            fontSize = 15.sp,
            fontWeight = FontWeight.SemiBold,
            color = if (isSelected) colors.accent else colors.ink,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
    }
}

// MARK: Деньги

/** Семантическая роль суммы — определяет цвет [MoneyText]. */
enum class MoneyRole {
    /** По знаку: `> 0` — accent, `< 0` — negative, `0` — inkSecondary. */
    AUTO,

    /** «Вам должны»/получено — всегда accent. */
    POSITIVE,

    /** «Вы должны»/долг — всегда negative. */
    NEGATIVE,

    /** Обычная сумма без семантики долга («Всего потрачено») — ink. */
    NEUTRAL,
}

/**
 * Единственный способ показать сумму: tabular-цифры (без «пляски» при
 * изменении), семантическая окраска. Всегда рисует МОДУЛЬ суммы — знак
 * передаётся цветом/контекстом (единая конвенция проекта).
 * Порт iOS `MoneyText`.
 */
@Composable
fun MoneyText(
    amount: Int,
    modifier: Modifier = Modifier,
    role: MoneyRole = MoneyRole.AUTO,
    size: TextUnit = 17.sp,
    weight: FontWeight = FontWeight.SemiBold,
    currency: String = "RUB",
) {
    val colors = Splitty.colors
    val color = when (role) {
        MoneyRole.POSITIVE -> colors.accent
        MoneyRole.NEGATIVE -> colors.negative
        MoneyRole.NEUTRAL -> colors.ink
        MoneyRole.AUTO -> when {
            amount > 0 -> colors.accent
            amount < 0 -> colors.negative
            else -> colors.inkSecondary
        }
    }
    Text(
        text = money(abs(amount), currency),
        modifier = modifier,
        color = color,
        fontSize = size,
        fontWeight = weight,
        maxLines = 1,
        style = androidx.compose.ui.text.TextStyle(
            fontFeatureSettings = "tnum", // моноширинные цифры
        ),
    )
}

/**
 * MoneyText с анимацией смены значения — аналог iOS `.contentTransition(
 * .numericText())`: при изменении [amount] старое число уезжает, новое
 * въезжает (кроссфейд+сдвиг). Гейтится reduce motion — тогда меняется мгновенно.
 * Для «живых» сумм (итог позиций чека, баланс после платежа).
 */
@Composable
fun AnimatedMoneyText(
    amount: Int,
    modifier: Modifier = Modifier,
    role: MoneyRole = MoneyRole.AUTO,
    size: TextUnit = 17.sp,
    weight: FontWeight = FontWeight.SemiBold,
    currency: String = "RUB",
) {
    val reduceMotion = rememberReduceMotion()
    if (reduceMotion) {
        MoneyText(amount, modifier, role, size, weight, currency)
        return
    }
    androidx.compose.animation.AnimatedContent(
        targetState = amount,
        modifier = modifier,
        transitionSpec = {
            val up = targetState > initialState
            val enter = androidx.compose.animation.fadeIn() +
                androidx.compose.animation.slideInVertically { if (up) it else -it }
            val exit = androidx.compose.animation.fadeOut() +
                androidx.compose.animation.slideOutVertically { if (up) -it else it }
            enter togetherWith exit
        },
        label = "numericMoney",
    ) { value ->
        MoneyText(value, role = role, size = size, weight = weight, currency = currency)
    }
}

/**
 * Итог, где могут встретиться РАЗНЫЕ валюты (общий баланс, нетто друга):
 * суммы не складываются между валютами — основная валюта (наибольший |суммы|)
 * крупно, остальные — вторичной строкой мельче через «·». [totals] — уже
 * агрегированные aggregateByCurrency (пустой список — «0 ₽» серым: расчёт).
 * Порт iOS `MoneyTotalsText`.
 */
@Composable
fun MoneyTotalsText(
    totals: List<CurrencySum>,
    modifier: Modifier = Modifier,
    primarySize: TextUnit = 40.sp,
    secondarySize: TextUnit = 15.sp,
    horizontalAlignment: Alignment.Horizontal = Alignment.Start,
) {
    Column(
        modifier = modifier,
        horizontalAlignment = horizontalAlignment,
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        val primary = totals.firstOrNull()
        if (primary != null) {
            MoneyText(primary.sum, size = primarySize, currency = primary.currency)
        } else {
            MoneyText(0, size = primarySize)
        }
        if (totals.size > 1) {
            Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                totals.drop(1).forEachIndexed { index, total ->
                    if (index > 0) {
                        Text(
                            text = "·",
                            fontSize = secondarySize,
                            fontWeight = FontWeight.SemiBold,
                            color = Splitty.colors.inkSecondary,
                        )
                    }
                    MoneyText(total.sum, size = secondarySize, currency = total.currency)
                }
            }
        }
    }
}

// MARK: Заголовок секции

/**
 * Заголовок секции: 13sp semibold, вторичный цвет, лёгкий kerning.
 * Регистр текста НЕ меняет. Аналог iOS `.sectionHeaderStyle()`.
 */
@Composable
fun SectionHeader(text: String, modifier: Modifier = Modifier, maxLines: Int = Int.MAX_VALUE) {
    Text(
        text = text,
        modifier = modifier,
        fontSize = 13.sp,
        fontWeight = FontWeight.SemiBold,
        color = Splitty.colors.inkSecondary,
        letterSpacing = 0.5.sp,
        maxLines = maxLines,
        overflow = TextOverflow.Ellipsis,
    )
}

File: ios/Splitty/Core/DesignSystem.swift
import SwiftUI
import UIKit

// MARK: - Масштабируемый шрифт (Dynamic Type)

/// Системный шрифт, масштабируемый настройкой размера текста: `size` —
/// базовый размер при стандартной настройке, растёт/уменьшается вместе с
/// `relativeTo`. Жёсткий `.font(.system(size:))` Dynamic Type игнорирует —
/// использовать этот модификатор для всех кастомных размеров.
private struct ScaledFontModifier: ViewModifier {
    @ScaledMetric private var size: CGFloat
    private let weight: Font.Weight
    private let design: Font.Design

    init(size: CGFloat, weight: Font.Weight, design: Font.Design, relativeTo style: Font.TextStyle) {
        _size = ScaledMetric(wrappedValue: size, relativeTo: style)
        self.weight = weight
        self.design = design
    }

    func body(content: Content) -> some View {
        content.font(.system(size: size, weight: weight, design: design))
    }
}

extension View {
    /// Шрифт с поддержкой Dynamic Type (см. `ScaledFontModifier`).
    func scaledFont(
        size: CGFloat,
        weight: Font.Weight = .regular,
        design: Font.Design = .rounded,
        relativeTo style: Font.TextStyle = .body
    ) -> some View {
        modifier(ScaledFontModifier(size: size, weight: weight, design: design, relativeTo: style))
    }
}

// MARK: - Карточка-поверхность

/// Премиум-карточка: фон `surface`, скругление 20pt (continuous);
/// светлая тема — мягкая тень, тёмная — hairline-бордер без тени.
struct SurfaceCardModifier: ViewModifier {
    var padding: CGFloat
    @Environment(\.colorScheme) private var colorScheme

    func body(content: Content) -> some View {
        let shape = RoundedRectangle(cornerRadius: 20, style: .continuous)
        content
            .padding(padding)
            .background(Color.surface, in: shape)
            .overlay {
                if colorScheme == .dark {
                    shape.strokeBorder(Color.hairline, lineWidth: 1)
                }
            }
            .shadow(
                color: colorScheme == .dark ? .clear : Color.black.opacity(0.06),
                radius: 14, x: 0, y: 6
            )
    }
}

extension View {
    /// Оформляет содержимое как мягкую карточку на фоне `Color.bg`.
    /// Использовать для секций-«карточек» вместо системных Form/List-фонов.
    func surfaceCard(padding: CGFloat = 16) -> some View {
        modifier(SurfaceCardModifier(padding: padding))
    }
}

// MARK: - Кнопки

/// Основной CTA: pill во всю ширину, высота 54pt, фон `accent`,
/// белый semibold-текст, pressed — scale 0.98 со spring-анимацией.
struct PrimaryPillButtonStyle: ButtonStyle {
    @Environment(\.isEnabled) private var isEnabled
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaledFont(size: 17, weight: .semibold)
            .foregroundStyle(.white)
            .frame(maxWidth: .infinity, minHeight: 54)
            .background(
                (configuration.isPressed ? Color.accentPressed : Color.accent)
                    .opacity(isEnabled ? 1 : 0.45),
                in: Capsule()
            )
            .scaleEffect(!reduceMotion && configuration.isPressed ? 0.98 : 1)
            .animation(.spring(duration: 0.25), value: configuration.isPressed)
    }
}

extension ButtonStyle where Self == PrimaryPillButtonStyle {
    /// Основной CTA-стиль («Сохранить», «Записать платёж», «Войти»…).
    static var primaryPill: PrimaryPillButtonStyle { PrimaryPillButtonStyle() }
}

/// Вторичная кнопка-чип: мягкая серая pill; `isSelected` — акцентная заливка
/// (для выбора группы/участника в чипах).
struct SoftChipButtonStyle: ButtonStyle {
    var isSelected: Bool = false
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaledFont(size: 15, weight: .semibold, relativeTo: .subheadline)
            // Чип не переносит текст по буквам в тесных строках («Погасить»).
            .lineLimit(1)
            .fixedSize(horizontal: true, vertical: false)
            .foregroundStyle(isSelected ? Color.accent : Color.ink)
            .padding(.horizontal, 16)
            .padding(.vertical, 10)
            .background(
                isSelected
                    ? Color.accent.opacity(0.14)
                    : Color.ink.opacity(configuration.isPressed ? 0.12 : 0.06),
                in: Capsule()
            )
            .scaleEffect(!reduceMotion && configuration.isPressed ? 0.98 : 1)
            .animation(.spring(duration: 0.25), value: configuration.isPressed)
    }
}

extension ButtonStyle where Self == SoftChipButtonStyle {
    /// Вторичная кнопка-чип («Балансы», «Итоги», фильтры…).
    static var softChip: SoftChipButtonStyle { SoftChipButtonStyle() }
    /// Чип с выбранным состоянием (акцентная заливка).
    static func softChip(isSelected: Bool) -> SoftChipButtonStyle {
        SoftChipButtonStyle(isSelected: isSelected)
    }
}

// MARK: - Деньги

/// Сумма в рублях по правилам премиум-дизайна: rounded + monospacedDigit,
/// семантическая окраска, `.contentTransition(.numericText())` при изменении.
/// Всегда показывает модуль суммы — знак передаётся цветом/контекстом
/// (единая конвенция проекта; единственная точка цветового правила денег).
struct MoneyText: View {
    /// Семантическая роль суммы (определяет цвет).
    enum Role {
        /// По знаку: `> 0` — accent, `< 0` — negative, `0` — inkSecondary.
        case auto
        /// «Вам должны»/получено — всегда accent.
        case positive
        /// «Вы должны»/долг — всегда negative.
        case negative
        /// Обычная сумма без семантики долга («Всего потрачено») — ink.
        case neutral
    }

    let amount: Int
    var role: Role = .auto
    var size: CGFloat = 17
    var weight: Font.Weight = .semibold
    /// Валюта суммы (код контракта); символ рисуется как у рублей: «1 234 $».
    var currency: String = "RUB"

    init(
        _ amount: Int,
        role: Role = .auto,
        size: CGFloat = 17,
        weight: Font.Weight = .semibold,
        currency: String = "RUB"
    ) {
        self.amount = amount
        self.role = role
        self.size = size
        self.weight = weight
        self.currency = currency
    }

    private var color: Color {
        switch role {
        case .positive: return .accent
        case .negative: return .negative
        case .neutral: return .ink
        case .auto:
            if amount > 0 { return .accent }
            if amount < 0 { return .negative }
            return .inkSecondary
        }
    }

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    var body: some View {
        Text(money(abs(amount), currency: currency))
            .scaledFont(size: size, weight: weight)
            .monospacedDigit()
            .foregroundStyle(color)
            .contentTransition(reduceMotion ? .identity : .numericText(value: Double(amount)))
            .animation(.spring(duration: 0.35), value: amount)
    }
}

// MARK: - Суммы в нескольких валютах

/// Итог, где могут встретиться РАЗНЫЕ валюты (общий баланс, нетто друга):
/// суммы не складываются между валютами — основная валюта (наибольший |суммы|)
/// крупно, остальные — вторичной строкой мельче. `totals` — уже агрегированные
/// `aggregateByCurrency` (пустой список — «0 ₽» серым: полный расчёт).
struct MoneyTotalsText: View {
    let totals: [CurrencySum]
    var primarySize: CGFloat = 40
    var secondarySize: CGFloat = 15
    var alignment: HorizontalAlignment = .leading

    var body: some View {
        VStack(alignment: alignment, spacing: 4) {
            if let primary = totals.first {
                MoneyText(primary.sum, size: primarySize, currency: primary.currency)
            } else {
                MoneyText(0, size: primarySize)
            }
            if totals.count > 1 {
                secondaryLine
            }
        }
    }

    /// Остальные валюты одной тихой строкой: «+120 $ · −3 400 Rp» мельче.
    private var secondaryLine: some View {
        HStack(spacing: 6) {
            ForEach(Array(totals.dropFirst().enumerated()), id: \.element) { index, total in
                if index > 0 {
                    Text("·")
                        .font(.system(size: secondarySize, weight: .semibold, design: .rounded))
                        .foregroundStyle(Color.inkSecondary)
                }
                MoneyText(total.sum, size: secondarySize, currency: total.currency)
            }
        }
    }
}

// MARK: - Заголовок секции

extension View {
    /// Стиль заголовка секции: 13pt semibold rounded, вторичный цвет.
    /// Регистр текста НЕ меняет (важно для UI-тестов).
    func sectionHeaderStyle() -> some View {
        self
            .scaledFont(size: 13, weight: .semibold, relativeTo: .footnote)
            .foregroundStyle(Color.inkSecondary)
            .kerning(0.5)
    }
}

// MARK: - Haptics

/// Хелперы тактильного отклика. Генераторы кэшированы: создание
/// UIFeedbackGenerator на каждое нажатие задерживало первый кадр после касания.
@MainActor
enum Haptics {
    private static let notificationGenerator = UINotificationFeedbackGenerator()
    private static let impactGenerator: UIImpactFeedbackGenerator = {
        let generator = UIImpactFeedbackGenerator(style: .light)
        generator.prepare()
        return generator
    }()

    /// Успешное сохранение/платёж/создание.
    static func success() {
        notificationGenerator.notificationOccurred(.success)
    }

    /// Лёгкий отклик на выбор/переключение (чипы, radio, чекбоксы).
    static func tap() {
        impactGenerator.impactOccurred()
        impactGenerator.prepare()
    }

    /// Действие недоступно/требует внимания (тап по заблокированной кнопке).
    static func warning() {
        notificationGenerator.notificationOccurred(.warning)
    }
}

// MARK: - Preview

#Preview("Дизайн-система") {
    ScrollView {
        VStack(spacing: 20) {
            VStack(alignment: .leading, spacing: 12) {
                Text("Общий баланс").sectionHeaderStyle()
                MoneyText(15200, size: 40)
                HStack {
                    MoneyText(-4300)
                    MoneyText(0)
                    MoneyText(990, role: .neutral)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .surfaceCard(padding: 20)

            Button("Записать платёж") {}
                .buttonStyle(.primaryPill)

            HStack {
                Button("Балансы") {}.buttonStyle(.softChip)
                Button("Итоги") {}.buttonStyle(.softChip(isSelected: true))
            }
        }
        .padding(20)
    }
    .background(Color.bg)
}
